### PR TITLE
[ Add ]  Utility methods of ParentNode & ChildNode interfaces

### LINKED ISSUE #288: [Bug] element.shadowRoot.append() is broken in Firefox
### Description
`element.shadowRoot.append( node )` will throw an error as shown below in Firefox:

> TypeError: 'append' called on an object that does not implement interface DocumentFragment

but `element.shadowRoot.appendChild( node )` works fine.

#### Live Demo
https://codepen.io/tech_query/pen/OBjzQr

### Versions
- webcomponents: v2.0.2+

### PR DESCRIPTION
**Pull request by TechQuery**
### Reference Issue

Fixes #288

### CODE REVIEW

**Review by TimvdLippe — CHANGES_REQUESTED** [review]
Please add braces to all if-statements. Also, please add regression tests for these methods.

**Comment by TechQuery** [comment]
@TimvdLippe 
I found **a bug of Building scripts**: only the name of [`prepend()`](https://github.com/EasyWebApp/shadydom/blob/51938be08b65991c2051ad6c030fd38d21ac4cfc/src/patch-builtins.js#L119) is transformed, I don't know why...
image

File: tests/parent-child-node.html [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    <script src="../node_modules/wct-browser-legacy/browser.js"></script>
+  </head>
+  <body>
+    <template>Text<!--Comment--><p>Element</p></template>
+
+    <div></div>
+
+    <script>
+      var childNodes,
```
Comment (TimvdLippe) [inline tests/parent-child-node.html]: We should define these variables as `let` and assign them in a `setup()`. Also, all tests should be wrapped in a `suite()`

Comment (TimvdLippe) [reply tests/parent-child-node.html]: You can see the other test suites for how that is done.

File: tests/parent-child-node.html [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+          return  node.nodeValue || node.textContent;
+        });
+      }
+
+      test('remove() of ChildNode interface',  function () {
+
+        childNodes = Array.apply(null, fragment.childNodes);
+
+        for (var i = 0;  childNodes[i];  i++)  childNodes[i].remove();
```
Comment (TimvdLippe) [inline tests/parent-child-node.html]: Always add braces for `for` and `if`

File: tests/parent-child-node.html [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+        assert.strictEqual(fragment.childNodes.length, 0);
+      });
+
+      test('append() of ParentNode interface',  function () {
+
+        sandBox.append.apply(sandBox, childNodes);
```
Comment (TimvdLippe) [inline tests/parent-child-node.html]: Just call `sandBox.append(childNodes)` here.

File: src/patch-builtins.js [hunk 4]
```diff
@@ -347,9 +415,11 @@ export function patchBuiltins() {
     HTMLElement;
   // These patches can always be done, for all supported browsers.
   patchBuiltin(ShadyRoot.prototype, shadowRootMixin);
+  patchBuiltin(ShadyRoot.prototype, interfaceParentNode);
   patchBuiltin(window.Node.prototype, nodeMixin);
   patchBuiltin(window.Window.prototype, windowMixin);
   patchBuiltin(window.Text.prototype, textMixin);
+  patchBuiltin(window.Comment.prototype, commentMixin);
```
Comment (TimvdLippe) [inline src/patch-builtins.js]: Nice!

File: src/patch-builtins.js [hunk 5]
```diff
@@ -17,6 +17,18 @@ import {addEventListener, removeEventListener} from './patch-events.js';
 import {attachShadow, ShadyRoot} from './attach-shadow.js';
 import {shadyDataForNode, ensureShadyDataForNode} from './shady-data.js';
 
+function fragmentFrom(list) {
+
+  const fragment = document.createDocumentFragment();
+
+  for (let node of list)
```
Comment (TimvdLippe) [inline src/patch-builtins.js]: Braces here as well.

File: package.json [hunk 6]
```diff
@@ -29,7 +29,7 @@
     "build": "gulp",
     "test": "npm run lint && npm run build && wct",
     "debug": "rollup -c",
-    "lint": "eslint src",
+    "lint": "eslint src --fix",
```
Comment (TimvdLippe) [inline package.json]: Please revert this change.

File: .editorconfig [hunk 7]
```diff
@@ -0,0 +1,9 @@
+root = true
```
Comment (TimvdLippe) [inline .editorconfig]: Please remove this file for now. We can always discuss this in a different PR.

File: src/patch-builtins.js [hunk 8]
```diff
@@ -135,10 +153,61 @@ let queryMixin = {
...
+/**
+ * @interface
+ *
+ * @see https://dom.spec.whatwg.org/#childnode
+ */
+let interfaceChildNode = {
+
+  before(...nodes) {
+    if (this.parentNode instanceof Node) {
+      this.parentNode.insertBefore(fragmentFrom( nodes ), this);
```
Comment (TimvdLippe) [inline src/patch-builtins.js]: Nit: unnecessary whitespace.

File: tests/parent-child-node.html [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+    <div></div>
+
+    <script>
+      var childNodes,
+        fragment = document.querySelector('template').content.cloneNode( true ),
+        sandBox = document.querySelector('div').attachShadow({mode: 'open'});
+
+      function getNodeText() {
+
```
Comment (TimvdLippe) [inline tests/parent-child-node.html]: Nit: Remove unnecessary whitespace.

**Review by TimvdLippe — CHANGES_REQUESTED** [review]
I think all methods need the following JSdoc notation:

```js
/**
  * @this {HTMLElement}
  */
```

This is also going to clash with #298, but I think it is better to wait for that PR to land before going ahead with this one.

File: tests/parent-child-node.html (hunk 2 above)
Comment (TechQuery) [reply tests/parent-child-node.html]: I think we should set a more strict **ESLint rule** to uniform code style of every coder.

File: tests/parent-child-node.html [hunk 10]
```diff
@@ -0,0 +1,87 @@
+<!DocType HTML>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <script src="../node_modules/@webcomponents/webcomponents-platform/webcomponents-platform.js"></script>
+    <script src="../node_modules/@webcomponents/template/template.js"></script>
+    <script>
+      ShadyDOM = {force: true};
+    </script>
+    <!-- <script src="../shadydom.min.js"></script> -->
```
[the remark this answers is not in the document]
Comment (TimvdLippe) [reply tests/parent-child-node.html]: Per my review comment, adding the following should fix that:

```js
/**
  * @this {HTMLElement}
  */
```

**Comment by dfreedm** [comment]
Sorry for the long delay.

If this issue is still important for you, could you please remake this PR in the new monorepo: https://github.com/webcomponents/polyfills?

Thanks!
